Track revisions

timkelty opened the pull request:
Adds feature to track revisions for deployed releases: https://github.com/shipitjs/shipit/issues/34

Tests are passing, but I didn't add any specific tests for this stuff yet.
Just wanted to get it out here first.

neoziro commented on lib/releases.js:
```diff
@@ -0,0 +1,110 @@
+ * @returns {string}
+ */
+
+var computeReleaseDirname = function computeReleaseDirname(result) {
```
You should use just:

```js
function computeReleaseDirname(result) {
... 
```

The `var computeReleaseDirname =` is useless.

neoziro replied:
Same for all methods.

timkelty replied:
No problem, I'll change.
At Fusionary, we tend to always use function expressions to [clarify hoisting and declaration order](https://javascriptweblog.wordpress.com/2010/07/06/function-declarations-vs-function-expressions/). Just a preference.

neoziro commented on lib/releases.js:
```diff
@@ -0,0 +1,110 @@
+  return shipit.remote('readlink ' + shipit.currentPath)
+  .then(function (results) {
+    if (!results)
+      return shipit.log('No current release found.');
```
I think you should return null here for consistency (null or a string, not undefined or whatever).

timkelty replied:
Done.

neoziro commented on lib/releases.js:
```diff
@@ -0,0 +1,110 @@
+  return shipit.remote('readlink ' + shipit.currentPath)
+  .then(function (results) {
+    if (!results) {
+      return shipit.log('No current release found.');
```
Same here, return null please.

timkelty replied:
Done.

neoziro commented on tasks/deploy/init.js:
```diff
@@ -1,5 +1,6 @@
 var registerTask = require('../../lib/register-task');
 var getShipit = require('../../lib/get-shipit');
+var path = require('path2/posix');
```
Why?

timkelty replied:
Ah - It was from this: https://github.com/timkelty/shipit-shared/pull/1, but I don't think we need it here.

neoziro commented on tasks/deploy/update.js:
```diff
@@ -15,8 +16,11 @@ module.exports = function (gruntOrShipit) {
 
   function task() {
     var shipit = getShipit(gruntOrShipit);
+    _.assign(shipit.constructor.prototype, require('../../lib/releases'));
```
You don't have to put that on the prototype, nobody will create a new Shipit object I think.

timkelty replied:
Ok.

neoziro commented on tasks/deploy/update.js:
```diff
@@ -50,5 +54,30 @@ module.exports = function (gruntOrShipit) {
+          var file = path.join(shipit.releasesPath, previousReleaseDir, 'REVISION');
+          return shipit.remote('if [ -f ' + file + ' ]; then cat ' + file + ' 2>/dev/null; fi;').then(function(response) {
+
+            // TODO: How should we handle multiple?
```
Multiple what?

timkelty replied:
`shipit.config.servers`. Should we just grab the result from the first, as I'm doing, assuming they should all be the same?

neoziro commented on tasks/deploy/update.js:
```diff
@@ -50,5 +54,30 @@ module.exports = function (gruntOrShipit) {
+    function setCurrentRevision() {
+      shipit.log('Setting current revision and creating revision file.');
+
+      return shipit.local('git rev-parse ' + shipit.config.branch).then(function(response) {
```
Just about coding style, for promise, put a break line before a new ".then", also try to avoid line with a length superior to ~ 200 chars (readability in editor and GitHub).

neoziro commented:
@timkelty OK except comments, and tests required please.

timkelty commented:
Ok, I'll review and make tests!
Thanks.

timkelty commented:
@neoziro OK, I think I've addressed everything you commented on, and I've added tests.
I added comments on your notes above.

Admittedly, I'm new to Mocha/Chai/Sinon, so let me know if I'm doing anything dumb with the tests, but they are passing.

neoziro commented on tasks/deploy/init.js:
```diff
@@ -1,5 +1,6 @@
 var registerTask = require('../../lib/register-task');
 var getShipit = require('../../lib/get-shipit');
+var path = require('path');
```
We must use path2 here.

timkelty commented:
Ok, I put `path2/posix` back in `init.js`.

neoziro commented on test/unit/tasks/deploy/update.js:
```diff
@@ -26,28 +28,120 @@ describe('deploy:update task', function () {
+      });
+    });
+
+    // it('should update remote REVISION file', function (done) {
```
Please remove commented code.

neoziro commented:
@timkelty last change and I merge

timkelty commented:
Ok, all set.
